Description: Fix SIXEL invalid colour register crash and remove SIXEL images before reflow.
Author: Nicholas Marriott <nicholas.marriott@gmail.com>
Origin: upstream, https://github.com/tmux/tmux/commit/aa17f0e0c1c8b3f1d6fc8617613c74f07de66fae
Bug: https://github.com/tmux/tmux/issues/3983
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1064363
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/tmux/+bug/2068393
Last-Update: 2024-07-10
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/image-sixel.c
+++ b/image-sixel.c
@@ -489,6 +489,9 @@
 		colours = si->colours;
 		ncolours = si->ncolours;
 	}
+
+	if (ncolours == 0)
+		return (NULL);
 	contains = xcalloc(1, ncolours);
 
 	len = 8192;
--- a/input.c
+++ b/input.c
@@ -1839,7 +1839,7 @@
 
 /* Handle CSI graphics SM. */
 static void
-input_csi_dispatch_sm_graphics(struct input_ctx *ictx)
+input_csi_dispatch_sm_graphics(__unused struct input_ctx *ictx)
 {
 #ifdef ENABLE_SIXEL
 	int	n, m, o;
--- a/screen-write.c
+++ b/screen-write.c
@@ -2283,6 +2283,10 @@
 		new = sixel_scale(si, 0, 0, 0, y - sy, sx, sy, 1);
 		sixel_free(si);
 		si = new;
+
+		/* Bail out if the image cannot be scaled. */
+		if (si == NULL)
+			return;
 		sixel_size_in_cells(si, &x, &y);
 	}
 
--- a/screen.c
+++ b/screen.c
@@ -308,12 +308,12 @@
 	if (sy != screen_size_y(s))
 		screen_resize_y(s, sy, eat_empty, &cy);
 
-	if (reflow) {
 #ifdef ENABLE_SIXEL
-		image_free_all(s);
+	image_free_all(s);
 #endif
+
+	if (reflow)
 		screen_reflow(s, sx, &cx, &cy, cursor);
-	}
 
 	if (cy >= s->grid->hsize) {
 		s->cx = cx;
